### ISSUE TITLE
Write a routine to read a subset of matrices from SuiteSparse (7 matrices) into KokkosSparse::CrsMatrix

### ISSUE THREAD

**Issue by uhetmaniuk** [fnrizzi/kokkos-kernels#5]
Using the matrices, [discussed here](https://github.com/fnrizzi/kokkos-kernels/issues/3#issuecomment-842759224), we need to read the files and convert them into CrsMatrix.
There is already a constructor to create a BlockCrsMatrix out of a CrsMatrix.
Then we can run some testing with the Mat-Vec product of Kokkos.

This reading capability may already be implemented in KokkosSparse::CrsMatrix.
In that case, we would need the driver for testing the time performance.

**Comment by mzuzek** [comment]
@uhetmaniuk

How would you decide on a block size ?
Since there is no inherent block size in the matrices from [the suite](https://sparse.tamu.edu), shall we just enforce on loaded matrix an arbitrary block size (passed as parameter) ?

**Comment by uhetmaniuk** [comment]
@MikolajZuzek 

I am thinking of using only the 7 matrices highlighted [here](https://github.com/fnrizzi/kokkos-kernels/issues/3#issuecomment-842759224).
For each of these 7 matrices, I would explicitly write in the code the block size.

Let me know if that doesn't answer your question.

**Comment by fnrizzi** [comment]
@uhetmaniuk yeah but how do you set/decide the block size? the block size should be passed to the construction of the block crs matrix. the matrices you point to are origianlly meant to be "regular crs matrices" so they don't have a native block size. 
I mean, yes, we can put block sIze = 4 for example, but that has no "meaning" in some sense. 
I agree that if we were to take for example a 2d finite volume code with multiple quantities per cell (e.g. density, u, v, temp) then the Jacobian matrix for such a system would have an inherent block size of 4. 
So all I am saying is that I agree that to test things practically we fix the block size to some number , but for those matrices linked above, the block size is undefined.

**Comment by uhetmaniuk** [comment]
@fnrizzi 

For the matrices that I linked, the column 3 is "bs = block size". 
So I am assuming that the authors of the paper used that block size value for the matrix.

The paper referenced in the [link ](https://github.com/fnrizzi/kokkos-kernels/issues/3#issuecomment-842759224) discusses the implementation of block compressed row storage.

**Comment by fnrizzi** [comment]
@uhetmaniuk sorry! I misread those, I actually thought you were referring to the gingko matrices. 
You are absolutely right, this makes perfect sense! (P.s. I know Mark the second author of that pdf )

**Comment by uhetmaniuk** [comment]
One option for these 7 matrices is to read them in KokkosSparse::CrsMatrix and then to use the constructor
```
  /// \brief Constructor that accepts a CrsMatrix and block dimension,
  ///        assuming the provided CrsMatrix has appropriate block structure.
  template<typename SType,
           typename OType,
           class DType,
           class MTType,
           typename IType>
  BlockCrsMatrix (const KokkosSparse::CrsMatrix<SType, OType, DType, MTType, IType> &crs_mtx,
                  const OrdinalType blockDimIn)
```

**Comment by fnrizzi** [comment]
yeah, makes perfect sense!

**Comment by uhetmaniuk** [comment]
@MikolajZuzek 

In the paper, they compute 3000 products and they report the average timings and flops per second.
So we could start by
- being able to do 3000 products when these 7 matrices are stored as `KokkosSparse::CrsMatrix`;
- reporting the results for serial, OpenMP, and GPU.

The PR #1 has a routine to do the matrix-vector product for KokkosSparse::Experimental::BlockCrsMatrix.
The `serial` case is pretty much done for the mat-vec product with `N` (normal). 

I need to finalize the `C` conjugate, the `T` transpose, and `H` the hermitian case.
I am looking now at the OpenMP version.

I can ping you when they are ready for testing.
Let me know if you need additional information for this current issue.

**Comment by uhetmaniuk** [comment]
LBV suggested to use the performance test

./perf_test/sparse/sparse_spmv

It looks like it can read matrices from the Matrix Market.
I have not tried it yet.

**Comment by uhetmaniuk** [comment]
@MikolajZuzek 

Please let me know when you have performance results for Kokkos with OpenMP.
Somehow on my Mac, OpenMP worsens the performance...

**Comment by fnrizzi** [comment]
@uhetmaniuk are you referring to this `./perf_test/sparse/sparse_spmv`? you mean running that with OpenMP is worse thanks serial?

**Comment by uhetmaniuk** [comment]
I was not able to run `sparse_spmv`... seems to take forever on my Mac.
I ran `./examples/wiki/sparse/KokkosKernels_nga_spmv` calling `spmv` with `KokkosSparse::CrsMatrix`
With OpenMP, the timings are lower.
I am suspecting that I am making a mistake somewhere.

**Comment by fnrizzi** [comment]
@uhetmaniuk
- this file `./examples/wiki/sparse/KokkosKernels_nga_spmv` does not exist, I guess you mean [https://github.com/fnrizzi/kokkos-kernels/blob/nga/create-SpMV-exe/example/wiki/sparse/KokkosSparse_nga_bspmv.cpp](https://github.com/fnrizzi/kokkos-kernels/blob/nga/create-SpMV-exe/example/wiki/sparse/KokkosSparse_nga_bspmv.cpp)? 
- so you tried to run `spmv` on line 337 with OpenMP, which picks up lines 506-536,  and this is slower than serial? 

Am I interpreting correctly?

**Comment by uhetmaniuk** [comment]
@fnrizzi 
* Yes, you are correct.
* No, I am talking about that [line](https://github.com/fnrizzi/kokkos-kernels/blob/94f5b090a706aaa5ec2d54640a9a1b483d777bce/example/wiki/sparse/KokkosSparse_nga_bspmv.cpp#L590)
It is the call to SpMV when the matrix is stored as CrsMatrix (even when the entries have a block structure).

**Comment by uhetmaniuk** [comment]
@fnrizzi 
I had my library built in Debug mode. The 'Release' mode gives now good results.

**Comment by fnrizzi** [comment]
That makes sense!

**Comment by mzuzek** [comment]
@uhetmaniuk @fnrizzi 

I've created draft PR #7 with `details::testSamples()` test routine that benchmarks matrices loaded from the samples (same way like generated matrices). KK has nice IO utils to load files in *.mtx and other formats, thanks for the pointer. Benchmark should be executed with our *.mtx files placed in current folder - with something like [get_samples.sh.txt](https://github.com/fnrizzi/kokkos-kernels/files/6574913/get_samples.sh.txt). 

I'm getting huge errors (`error` ~= `maxNorm`) for the samples: block and reference `spmv` yield completely different results... Don't know why yet, will be diagnosing that (errors from the randomized routine are still fine).

**Comment by mzuzek** [comment]
After serializing and deserializing the matrix generated with `details::generate_crs_matrix()` (with `blockSize` 12) we get decent errors around 1.8e-14, which would indicate that crs matrix serialization and deserialization works fine (`KokkosKernels::Impl::write_kokkos_crst_matrix()` and `KokkosKernels::Impl::read_kokkos_crst_matrix()` respectively). When some values were expanded to range around [-13e+3, 13e+3] (like in _GT01R_) the error increased to about 4.6e-10.

I guess we need to fix that before timing the samples ?  
@uhetmaniuk Would you be able to check why KK spmv and our block version give completely different results on sample matrices (i.e. when you run `testSamples()`) ?  
Let me know where I can help.

**Comment by uhetmaniuk** [comment]
@MikolajZuzek 

I have fixed the construction of the BlockCrsMatrix. The current constructors in Kokkos assume that the number of rows and the number of columns are multiple of the block size and that the number of non-zero entries is a multiple of the block size squared.

My fix seems to be good for GT01R and audikw_1.  We may have to skip the matrices bmw7st_1 and pwtk. 
There is still an error for the matrices raefsky4 and RM07R. 
It could be inside my constructor but I did not have time to check further.
